Python/Tests/Core/WorkspaceInterpreterFactoryTests.cs

```diff
@@ -159,15 +159,74 @@ public class WorkspaceInterpreterFactoryTests {
             }
         }
 
-        private static IEnumerable<InterpreterConfiguration> TestTriggerDiscovery(IVsFolderWorkspaceService workspaceService, Action triggerDiscovery) {
+        [TestMethod, Priority(0)]
+        public void WatchWorkspaceVirtualEnvRenamed() {
+            const string ENV_NAME = "env";
+            var workspaceService = CreateEnvAndGetService(ENV_NAME);
+
+            string envPath = Path.Combine(workspaceService.CurrentWorkspace.Location, ENV_NAME);
+            var configs = TestTriggerDiscovery
+                (workspaceService, () => Directory.Move(envPath, string.Concat(envPath, "1"))).ToArray();
+
+            Assert.AreEqual(1, configs.Length);
+            Assert.IsTrue(PathUtils.IsSamePath(
+                Path.Combine(string.Concat(envPath, "1"), "scripts", "python.exe"),
+                configs[0].InterpreterPath));
+            Assert.AreEqual("Workspace|Workspace|env1", configs[0].Id);
+
+            Directory.Delete(workspaceService.CurrentWorkspace.Location, true);
+        }
+
+        [TestMethod, Priority(0)]
+        public void WatchWorkspaceVirtualEnvDeleted() {
+            const string ENV_NAME = "env";
+            var workspaceService = CreateEnvAndGetService(ENV_NAME);
+
+            string envPath = Path.Combine(workspaceService.CurrentWorkspace.Location, ENV_NAME);
+            var configs = TestTriggerDiscovery(workspaceService, () => Directory.Delete(envPath, true)).ToArray();
+
+            Assert.AreEqual(0, configs.Length);
+        }
+
+        private MockWorkspaceService CreateEnvAndGetService(string envName) {
+            var python = PythonPaths.Python37_x64 ?? PythonPaths.Python37;
+
+            var workspacePath = TestData.GetTempPath();
+            Directory.CreateDirectory(workspacePath);
+            File.WriteAllText(Path.Combine(workspacePath, "app.py"), string.Empty);
+
+            //Creating virtual environment and confirming it was created
+            using (var p = ProcessOutput.RunHiddenAndCapture(python.InterpreterPath, "-m", "venv", Path.Combine(workspacePath, envName))) {
+                Console.WriteLine(p.Arguments);
+                Assert.IsTrue(p.Wait(TimeSpan.FromMinutes(3)));
+                Console.WriteLine(string.Join(Environment.NewLine, p.StandardOutputLines.Concat(p.StandardErrorLines)));
+                Assert.AreEqual(0, p.ExitCode);
+            }
+            Assert.IsTrue(File.Exists(Path.Combine(workspacePath, envName, "scripts", "python.exe")));
+
```

Comment:
this creating virtual env block 198-205) is worth extracting into its own method and reuse it in a couple of places.

Reply:
Refactored